File: C#/Loon2MonoGame/LoonMonoGame-Lib/loon/opengl/Painter.cs

using loon.geom;
using loon.utils;
using loon.utils.reply;

namespace loon.opengl
{
  public  abstract class Painter : TextureSource
	{
		public object Tag;

		public const int TOP_LEFT = 0;

		public const int TOP_RIGHT = 1;

		public const int BOTTOM_RIGHT = 2;

		public const int BOTTOM_LEFT = 3;

		public abstract LTexture Texture();

		public abstract float Width();

		public abstract float Height();

		public abstract float GetDisplayWidth();

		public abstract float GetDisplayHeight();

		public abstract float Sx();

		public abstract float Sy();

		public abstract float Tx();

		public abstract float Ty();

		public abstract void AddToBatch(BaseBatch batch, uint tint, Affine2f tx, float x, float y, float width, float height);

		public abstract void AddToBatch(BaseBatch batch, uint tint, Affine2f tx, float dx, float dy, float dw, float dh, float sx, float sy, float sw, float sh);

		public override bool IsLoaded()
		{
		
				return _isLoaded;
			
		}

		public override Painter Draw()
		{
			return this;
		}

		public override GoFuture<Painter> TileAsync()
		{
			return GoFuture<Painter>.Success(this);
		}

		public override string ToString()
		{
			StringKeyValue builder = new StringKeyValue("Painter");
			builder.Kv("size", Width() + "x" + Height()).Comma().Kv("xOff", Sx()).Comma().Kv("yOff", Sy()).Comma().Kv("widthRatio", Tx()).Comma().Kv("heightRatio", Ty());
			return builder.ToString() + " <- " + Texture();
		}
	}
}

File: Java/Loon-Neo/src/loon/opengl/Painter.java
/**
 * Copyright 2008 - 2015 The Loon Game Engine Authors
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.5
 */
package loon.opengl;

import loon.LTexture;
import loon.geom.Affine2f;
import loon.utils.StringKeyValue;
import loon.utils.reply.GoFuture;

public abstract class Painter extends TextureSource {

	public Object Tag;

	public static final int TOP_LEFT = 0;

	public static final int TOP_RIGHT = 1;

	public static final int BOTTOM_RIGHT = 2;

	public static final int BOTTOM_LEFT = 3;

	public static LTexture firstFather(LTexture texture) {
		if (texture.getParent() == null) {
			return texture;
		} else {
			return firstFather(texture.getParent());
		}
	}

	public abstract LTexture texture();

	public abstract float width();

	public abstract float height();

	public abstract float getDisplayWidth();

	public abstract float getDisplayHeight();

	public abstract float sx();

	public abstract float sy();

	public abstract float tx();

	public abstract float ty();

	public abstract void addToBatch(BaseBatch batch, int tint, Affine2f tx, float x, float y, float width,
			float height);

	public abstract void addToBatch(BaseBatch batch, int tint, Affine2f tx, float dx, float dy, float dw, float dh,
			float sx, float sy, float sw, float sh);

	public abstract void quad(BaseBatch batch, int tint, Affine2f tx, float x1, float y1, float x2, float y2, float x3,
			float y3, float x4, float y4);

	@Override
	public boolean isLoaded() {
		return _isLoaded;
	}

	@Override
	public Painter draw() {
		return this;
	}

	@Override
	public GoFuture<Painter> tileAsync() {
		return GoFuture.success(this);
	}

	@Override
	public String toString() {
		StringKeyValue builder = new StringKeyValue("Painter");
		builder.kv("size", width() + "x" + height()).comma().kv("xOff", sx()).comma().kv("yOff", sy()).comma()
				.kv("widthRatio", tx()).comma().kv("heightRatio", ty());
		return builder.toString() + " <- " + texture();
	}
}
